modules/gapi/src/backends/fluid/gfluidcore_func.simd.hpp

```diff
@@ -821,6 +845,322 @@ MUL_SIMD(float, float)
+                              v_pack(v_round(vg_load_f32(&in[x + 5*nlanes/2]) + s2),
+                                     v_round(vg_load_f32(&in[x + 11*nlanes/4]) + s3))));
+        }
+
+        if (x < length)
+        {
+            x = length - lanes;
+            continue;  // process unaligned tail
+        }
+        break;
+    }
+    return x;
+}
+
+//-------------------------------------------------------------------------------------------------
+
+template<typename SRC>
+CV_ALWAYS_INLINE int addc_simd_c3_impl(const SRC in[], float out[],
+                                       const v_float32& s1, const v_float32& s2,
+                                       const v_float32& s3, const int length)
+{
+    constexpr int nlanes = v_float32::nlanes;
+    constexpr int chan = 3;
+    constexpr int lanes = chan * nlanes;
+
+    int x = 0;
+    for (;;)
+    {
+        for (; x <= length - lanes; x += lanes)
+        {
+            v_float32 a1 = vg_load_f32(&in[x]);
+            v_float32 a2 = vg_load_f32(&in[x + nlanes]);
+            v_float32 a3 = vg_load_f32(&in[x + 2*nlanes]);
+
+            vx_store(&out[x], a1 + s1);
+            vx_store(&out[x + nlanes], a2 + s2);
+            vx_store(&out[x + 2*nlanes], a3 + s3);
+        }
+
+        if (x < length)
+        {
+            x = length - lanes;
+            continue;  // process unaligned tail
+        }
+        break;
+    }
+    return x;
+}
+
+//-------------------------------------------------------------------------------------------------
+
+template<typename SRC, typename DST>
+CV_ALWAYS_INLINE int addc_simd_c3(const SRC in[], const float scalar[], DST out[], const int length)
+{
+    constexpr int chan = 3;
+    constexpr int nlanes = vector_type_of_t<DST>::nlanes;
+    constexpr int lanes = chan * nlanes;
+
+    if (length < lanes)
+        return 0;
+
+    v_float32 s1 = vx_load(scalar);
+#if CV_SIMD_WIDTH == 32
+    v_float32 s2 = vx_load(&scalar[2]);
+    v_float32 s3 = vx_load(&scalar[1]);
+#else
+    v_float32 s2 = vx_load(&scalar[1]);
+    v_float32 s3 = vx_load(&scalar[2]);
+#endif
+
+    return addc_simd_c3_impl(in, out, s1, s2, s3, length);
+}
+
+#define ADDC_SIMD(SRC, DST)                                       \
+int addc_simd(const SRC in[], const float scalar[], DST out[],    \
+              const int width, const int chan)                    \
+{                                                                 \
+    const int length = width * chan;                              \
+    switch (chan)                                                 \
+    {                                                             \
+    case 1:                                                       \
+    case 2:                                                       \
+    case 4:                                                       \
+        return addc_simd_impl(in, scalar, out, length);           \
```

Comment:
Why 1, 2 & 4 falls into the same case

Reply:
It is related to data alignment. Scalar array's coefficients for the whole number of pixels fit into one SIMD vector.
**SSE(128 bit) example:**
1-channel picture:
Scalar SIMD vector contains: 
C1 C1 C1 C1

2-channel picture:
Scalar SIMD vector contains:
C1 C2 C1 C2

4-channel picture:
Scalar SIMD vector contains:
C1 C2 C3 C4

**_Exception is 3-channel picture:_**
First scalar SIMD vector contains:
C1 C2 C3 **_C1_**